app/app.go

```diff
@@ -856,12 +865,87 @@ func (app *App) setPostHandler() {
 	app.SetPostHandler(postHandler)
 }
 
+func (app *App) setAdmissionHandler(handler consensus.AdmissionHandler) {
+	app.consensusMessageAdmissionHandler = handler
+}
+
 // Name returns the name of the App
 func (app *App) Name() string { return app.BaseApp.Name() }
 
 // BeginBlocker application updates every begin block
 func (app *App) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) abci.ResponseBeginBlock {
-	return app.mm.BeginBlock(ctx, req)
+	consensusResponses := app.processConsensusMessage(ctx, req.ConsensusMessages)
+
+	resp := app.mm.BeginBlock(ctx, req)
+	resp.ConsensusMessagesResponses = consensusResponses
+
+	return resp
+}
+
+func (app *App) processConsensusMessage(ctx sdk.Context, consensusMsgs []*prototypes.Any) []*abci.ConsensusMessageResponse {
+	var responses []*abci.ConsensusMessageResponse
+
+	for _, anyMsg := range consensusMsgs {
+		sdkAny := &types.Any{
+			TypeUrl: "/" + anyMsg.TypeUrl,
+			Value:   anyMsg.Value,
+		}
+
+		var msg sdk.Msg
+		err := app.appCodec.UnpackAny(sdkAny, &msg)
+		if err != nil {
+			responses = append(responses, &abci.ConsensusMessageResponse{
+				Response: &abci.ConsensusMessageResponse_Error{
+					Error: fmt.Errorf("failed to unpack consensus message: %w", err).Error(),
+				},
+			})
+
+			continue
+		}
+
+		cacheCtx, writeCache := ctx.CacheContext()
+		err = app.consensusMessageAdmissionHandler(cacheCtx, msg)
+		if err != nil {
+			responses = append(responses, &abci.ConsensusMessageResponse{
+				Response: &abci.ConsensusMessageResponse_Error{
+					Error: fmt.Errorf("consensus message admission failed: %w", err).Error(),
+				},
+			})
+
+			continue
+		}
+
+		resp, err := app.MsgServiceRouter().Handler(msg)(ctx, msg)
```

Comment:
shouldn't we here pass cache ctx and than abort in case of errors?
deliver tx provides it for us automatically but afair in begin block we need to handle it manually.
i.e - if the handler fails, how does the state changes get reverted?